### PR TITLE
Thread assembly signing through CI pipelines and add signed IVT for extension projects

### PR DESCRIPTION
**Pull request by paulmedynski**
## Summary

Closes the assembly signing gap in CI pipelines for Abstractions, Azure, and Logging extension projects. After this change, `#if ASSEMBLY_SIGNING` code paths are exercised during CI test runs for all extension packages, and test assemblies can access internals from signed source assemblies via properly keyed `InternalsVisibleTo` grants.

This was driven by the missing assembly public key checks in Abstrations SqlAuthenticationProvider.Internal, and I discovered we aren't actually signing and testing those paths in our CI pipelines.

### Pipeline infrastructure

- Create reusable `eng/pipelines/common/steps/download-assembly-signing-key.yml` template with `isTest` parameter for driver vs test key download
- Migrate all pack jobs to use the template (Abstractions, Azure, Logging, SqlClient, OneBranch)
- Add `isInternalBuild` concept to the CI pipeline so it can decide when to sign assemblies which allows us to select:
  - When InternalsVisibleTo is granted to test projects, and when those grants require the test signing public key.
  - When ASSEMBLY_SIGNING is defined during builds, and which conditional code blocks to compile.
- Add test signing key to the build and pipeline flows that deal with tests that must be signed in order to access their associated driver project IVT grants.

### Source and project file changes

- Add `ASSEMBLY_SIGNING` conditional compilation to `SqlAuthenticationProvider.Internal.cs`
  - This is the main functional driver change that all of the other changes support.
- Add dual `InternalsVisibleTo` blocks to Abstractions, Logging, and Azure source projects:
  - Unsigned (`SigningKeyPath=''`): grants access to test assembly
  - Signed + Package mode (`SigningKeyPath != '' AND ReferenceType == 'Package'`): grants access with test public key
- Wire up the `isInternalBuild`, `SigningKeyPath`, and `TestSigningKeyPath` through the pipeline, build, and project files.

### Logging.Test project and CI coverage

- Create `Logging.Test.csproj` targeting `net462;net8.0;net9.0;net10.0`
- Add `SqlClientEventSourceTest` verifying the singleton instance
- Add PR/CI pipeline support for this new test project.

This avoids the absence of this test project causing the build and pipeline plumbing from existing, which was a confusing divergence from the other changes here.  Also, this project _should_ have some tests, and this creates the empty project where we can easily add tests in a future PR.

### Rename STRONG_NAME_SIGNING to ASSEMBLY_SIGNING

- The new name better reflects what the build/test/pipeline environment is actually doing across TFMs.
  - .NET only does basic public-key signing; it does not perform strong-name signing like .NET Framework does.

### Disable signing in Project-mode pipelines

- Gate all signing key downloads and `SigningKeyPath` propagation on `ne(referenceType, 'Project')` across:
  - Pack jobs: Abstractions, Azure, Logging, SqlServer.Server
  - Test jobs: Abstractions, Azure, Logging
  - Common templates: `ci-run-tests-job.yml`, `ci-build-nugets-job.yml`
- Forward `referenceType` parameter to leaf package stages/jobs that previously lacked it (Logging, SqlServer.Server, Abstractions test jobs)
- This ensures Project-mode builds remain unsigned (preserving IVT without public key), while Package-mode builds are signed for real package consumption.

## Testing

- [x] `dotnet build build.proj -t:TestAbstractions` — 29 tests pass (net8.0/net9.0)
- [x] `dotnet build build.proj -t:TestLogging` — 1 test passes
- [x] `dotnet build build.proj -t:TestAzure` — 4 pass, 16 skipped (no SQL Server)
- [x] All modified YAML files pass `yaml.safe_load()` validation
- [x] Package-mode pipeline (build 157028): all 20 assemblies across 6 packages verified signed with PEReader
- [x] Project-mode pipeline (build 157031): SqlClient + AKV assemblies verified unsigned

## Design Notes

- Signed IVT is intentionally limited to **Package mode only**. In Project mode, source is compiled alongside tests so signing is not needed for IVT.
- `signingArguments` pipeline variable is always defined (empty string for non-internal builds) to avoid undefined `$(...)` variable expansion issues in ADO.
- Key downloads are gated on both `isInternalBuild` and `ne(referenceType, 'Project')` — this avoids downloading keys that won't be used.

### CODE REVIEW

File: src/Microsoft.Data.SqlClient.Internal/Logging/test/Logging.Test.csproj [hunk 1]
```diff
@@ -0,0 +1,51 @@
+<Project Sdk="Microsoft.NET.Sdk">
+  <!-- General Properties ============================================== -->
+  <PropertyGroup>
+    <AssemblyName>Microsoft.Data.SqlClient.Internal.Logging.Test</AssemblyName>
+    <TargetFrameworks>net462;net8.0;net9.0;net10.0</TargetFrameworks>
+
```
[the remark this answers is not in the document]
Comment (paulmedynski) [reply src/Microsoft.Data.SqlClient.Internal/Logging/test/Logging.Test.csproj:12]: Fixed — made `net462` conditional on `$(OS) == 'Windows_NT'`, matching the pattern in Azure.Test.csproj. Cross-platform jobs will now only build/test `net8.0`/`net9.0`/`net10.0`.

File: eng/pipelines/jobs/test-logging-package-ci-job.yml [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+      # Building this project implicitly builds the Logging project.
+      - name: project
+        value: src/Microsoft.Data.SqlClient.Internal/Logging/test/Logging.Test.csproj
+
+      # dotnet CLI arguments for build/test/pack commands
+      - name: buildArguments
+        value: >-
+          -p:Configuration=${{ parameters.buildConfiguration }}
+          --verbosity ${{ parameters.dotnetVerbosity }}
+
```
[the remark this answers is not in the document]
Comment (paulmedynski) [reply eng/pipelines/jobs/test-logging-package-ci-job.yml:110]: Fixed — renamed `buildArguments` to `dotnetBuildOpts` to avoid the `{command}ARGUMENTS` implicit injection pattern.

File: eng/pipelines/jobs/test-abstractions-package-ci-job.yml [hunk 3]
```diff
@@ -136,7 +159,7 @@ jobs:
         inputs:
           command: build
           projects: $(project)
-          arguments: $(buildArguments)
+          arguments: $(buildArguments) $(signingArguments)
```
[the remark this answers is not in the document]
Comment (paulmedynski) [reply eng/pipelines/jobs/test-abstractions-package-ci-job.yml:171]: Fixed — renamed `buildArguments` to `dotnetBuildOpts` here as well.

File: eng/pipelines/jobs/test-azure-package-ci-job.yml [hunk 4]
```diff
@@ -289,7 +312,7 @@ jobs:
         inputs:
           command: build
           projects: $(project)
-          arguments: $(buildArguments)
+          arguments: $(buildArguments) $(signingArguments)
```
[the remark this answers is not in the document]
Comment (paulmedynski) [reply eng/pipelines/jobs/test-azure-package-ci-job.yml:316]: Fixed — renamed `buildArguments` to `dotnetBuildOpts` here as well.

File: src/Microsoft.Data.SqlClient.Internal/Logging/test/SqlClientEventSourceTest.cs [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+namespace Microsoft.Data.SqlClient.Internal.Logging.Test;
+
+public class SqlClientEventSourceTest
+{
+    [Fact]
+    public void SqlClientEventSource_Log_IsNotNull()
+    {
+        Assert.NotNull(SqlClientEventSource.Log);
+    }
+}
```
[the remark this answers is not in the document]
Comment (paulmedynski) [reply src/Microsoft.Data.SqlClient.Internal/Logging/test/SqlClientEventSourceTest.cs:20]: Fixed — added a separate `SqlClientEventSource_Log_IsSingleton` test using `Assert.Same` to verify singleton identity. Kept the original `Assert.NotNull` test as well.

File: src/Microsoft.Data.SqlClient.Extensions/Abstractions/src/SqlAuthenticationProvider.Internal.cs [hunk 6]
```diff
@@ -23,6 +23,14 @@ public abstract partial class SqlAuthenticationProvider
...
     private static class Internal
     {
+        /// <summary>
+        /// The expected public key token of the SqlClient assembly, used to avoid loading imposter
+        /// assemblies.  This is the public key token of the assembly signing key used for all of
+        /// our driver assemblies.
+        /// </summary>
+        private static readonly byte[] _sqlClientPublicKeyToken =
+            [ 0x23, 0xec, 0x7f, 0xc2, 0xd6, 0xea, 0xa4, 0xa5 ];
+
```
[the remark this answers is not in the document]
Comment (paulmedynski) [reply src/Microsoft.Data.SqlClient.Extensions/Abstractions/src/SqlAuthenticationProvider.Internal.cs]: Fixed in 240ea7e57. Replaced the static `_sqlClientPublicKeyToken` field with a local `expectedPublicKeyToken` variable inside the `#if ASSEMBLY_SIGNING` block, eliminating the CS0414 warning in unsigned builds.

File: src/Microsoft.Data.SqlClient.Extensions/Abstractions/test/Abstractions.Test.csproj [hunk 7]
```diff
@@ -8,6 +8,13 @@
     <IsTestProject>true</IsTestProject>
   </PropertyGroup>
 
+  <!-- Assembly signing ============================================= -->
+  <!-- When a test signing key is provided, sign the test assembly so IVT from signed source works. -->
+  <PropertyGroup Condition="'$(TestSigningKeyPath)' != ''">
+    <SignAssembly>true</SignAssembly>
+    <AssemblyOriginatorKeyFile>$(TestSigningKeyPath)</AssemblyOriginatorKeyFile>
```
[the remark this answers is not in the document]
Comment (paulmedynski) [reply src/Microsoft.Data.SqlClient.Extensions/Abstractions/test/Abstractions.Test.csproj:22]: Fixed in 240ea7e57. Aligned with sibling test projects (Logging.Test, Azure.Test) by making `net462` conditional on Windows via `Condition="'$(OS)' == 'Windows_NT'"`, with an explanatory comment.

File: eng/pipelines/jobs/pack-azure-package-ci-job.yml [hunk 8]
```diff
@@ -157,32 +181,20 @@ jobs:
         parameters:
           debug: ${{ parameters.debug }}
 
+      # Download the assembly signing key for internal builds.
+      - ${{ if eq(parameters.isInternalBuild, true) }}:
+        - template: /eng/pipelines/common/steps/download-assembly-signing-key.yml@self
+
```
[the remark this answers is not in the document]
Comment (paulmedynski) [reply eng/pipelines/jobs/pack-azure-package-ci-job.yml]: Fixed in 8668069de. Key download in `pack-azure-package-ci-job.yml` is now gated on `ne(parameters.referenceType, 'Project')`.

File: eng/pipelines/onebranch/jobs/validate-signed-package-job.yml [hunk 9]
```diff
@@ -157,8 +157,8 @@ jobs:
           $nugetPackageInstallPath = "${{ variables.nugetPackageInstallPath }}"
           echo "nugetPackageInstallPath= $nugetPackageInstallPath"
 
-          # Verify strong name signing #####################################
-          echo "> 1. Verifying strong name signing of DLLs ..."
+          # Verify assembly signing #####################################
+          echo "> 1. Verifying assembly signing of DLLs ..."
```
[the remark this answers is not in the document]
Comment (paulmedynski) [reply eng/pipelines/onebranch/jobs/validate-signed-package-job.yml]: Fixed in 240ea7e57. Changed terminology from "assembly signing" to "strong-name signing" in both the comment and echo text.

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlAuthenticationProviderManager.cs [hunk 10]
```diff
@@ -109,7 +109,7 @@ static SqlAuthenticationProviderManager()
                 SqlClientEventSource.Log.TryTraceEvent(
                     nameof(SqlAuthenticationProviderManager) +
                     $": Attempting to load Azure extension assembly={azureAssemblyName} without " +
-                    "strong name verification; ensure this assembly is from a trusted source");
+                    "assembly verification; ensure this assembly is from a trusted source");
```
[the remark this answers is not in the document]
Comment (paulmedynski) [reply src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlAuthenticationProviderManager.cs:113]: Fixed in 240ea7e57. Changed "without assembly verification" to "without strong-name identity verification" in the trace message.

File: src/Microsoft.Data.SqlClient.Extensions/Abstractions/src/SqlAuthenticationProvider.Internal.cs [hunk 11]
```diff
@@ -23,6 +23,14 @@ public abstract partial class SqlAuthenticationProvider
     /// </summary>
     private static class Internal
     {
+        /// <summary>
+        /// The expected public key token of the SqlClient assembly, used to avoid loading imposter
+        /// assemblies.  This is the public key token of the assembly signing key used for all of
+        /// our driver assemblies.
+        /// </summary>
```
[the remark this answers is not in the document]
Comment (paulmedynski) [reply src/Microsoft.Data.SqlClient.Extensions/Abstractions/src/SqlAuthenticationProvider.Internal.cs]: Fixed in 240ea7e57. Rephrased from "loading imposter assemblies" to "invoking APIs from imposter assemblies" to accurately reflect the protection mechanism (the check happens after loading, before invoking reflection APIs).

File: eng/pipelines/jobs/test-azure-package-ci-job.yml [hunk 12]
```diff
@@ -179,6 +185,16 @@ jobs:
...
+      # Signing arguments — only set for internal builds.
+      - ${{ if eq(parameters.isInternalBuild, true) }}:
+        - name: signingArguments
+          value: >-
+            -p:SigningKeyPath=$(driverKeyFile.secureFilePath)
+            -p:TestSigningKeyPath=$(testKeyFile.secureFilePath)
+      - ${{ else }}:
+        - name: signingArguments
+          value: ''
+
```
[the remark this answers is not in the document]
Comment (paulmedynski) [reply eng/pipelines/jobs/test-azure-package-ci-job.yml]: Fixed in 5b5ff1216. Signing in `test-azure-package-ci-job.yml` is now gated on `ne(parameters.referenceType, 'Project')`.

File: eng/pipelines/jobs/test-azure-package-ci-job.yml [hunk 13]
```diff
@@ -206,6 +222,13 @@ jobs:
         - pwsh: 'Get-ChildItem Env: | Sort-Object Name'
           displayName: '[Debug] Print Environment Variables'
 
+      # Download the assembly signing keys for internal builds.
+      - ${{ if eq(parameters.isInternalBuild, true) }}:
+        - template: /eng/pipelines/common/steps/download-assembly-signing-key.yml@self
+        - template: /eng/pipelines/common/steps/download-assembly-signing-key.yml@self
+          parameters:
+            isTest: true
+
```
[the remark this answers is not in the document]
Comment (paulmedynski) [reply eng/pipelines/jobs/test-azure-package-ci-job.yml]: Fixed in 5b5ff1216. Key download in `test-azure-package-ci-job.yml` is now gated on `ne(parameters.referenceType, 'Project')`.

File: eng/pipelines/jobs/test-abstractions-package-ci-job.yml [hunk 14]
```diff
@@ -89,11 +95,21 @@ jobs:
...
 
+      # Signing arguments — only set for internal builds.
+      - ${{ if eq(parameters.isInternalBuild, true) }}:
+        - name: signingArguments
+          value: >-
+            -p:SigningKeyPath=$(driverKeyFile.secureFilePath)
+            -p:TestSigningKeyPath=$(testKeyFile.secureFilePath)
+      - ${{ else }}:
+        - name: signingArguments
+          value: ''
```
[the remark this answers is not in the document]
Comment (paulmedynski) [reply eng/pipelines/jobs/test-abstractions-package-ci-job.yml]: Fixed in 5b5ff1216. Added `referenceType` parameter to `test-abstractions-package-ci-job.yml` and gated signing on `ne(parameters.referenceType, 'Project')`.

File: eng/pipelines/jobs/test-logging-package-ci-job.yml [hunk 15]
```diff
@@ -0,0 +1,210 @@
...
+      # Signing arguments — only set for internal builds.
+      - ${{ if eq(parameters.isInternalBuild, true) }}:
+        - name: signingArguments
+          value: >-
+            -p:SigningKeyPath=$(driverKeyFile.secureFilePath)
+            -p:TestSigningKeyPath=$(testKeyFile.secureFilePath)
+      - ${{ else }}:
+        - name: signingArguments
+          value: ''
```
[the remark this answers is not in the document]
Comment (paulmedynski) [reply eng/pipelines/jobs/test-logging-package-ci-job.yml]: Fixed in 5b5ff1216. Added `referenceType` parameter to `test-logging-package-ci-job.yml` and gated signing on `ne(parameters.referenceType, 'Project')`.

File: eng/pipelines/common/templates/jobs/ci-run-tests-job.yml [hunk 16]
```diff
@@ -223,6 +228,13 @@ jobs:
   # Restore dotnet CLI tools (e.g. pwsh, apicompat) before building.
   - template: /eng/pipelines/common/steps/restore-dotnet-tools.yml@self
 
+  # Download the assembly signing keys for internal builds.
+  - ${{ if eq(parameters.isInternalBuild, true) }}:
+    - template: /eng/pipelines/common/steps/download-assembly-signing-key.yml@self
+    - template: /eng/pipelines/common/steps/download-assembly-signing-key.yml@self
+      parameters:
+        isTest: true
+
```
[the remark this answers is not in the document]
Comment (paulmedynski) [reply eng/pipelines/common/templates/jobs/ci-run-tests-job.yml]: Fixed in 240ea7e57. Key download steps in `ci-run-tests-job.yml` are now gated on `ne(referenceType, 'Project')` in addition to `isInternalBuild`, so keys are only downloaded when signing is active.

File: eng/pipelines/common/templates/jobs/ci-build-nugets-job.yml [hunk 17]
```diff
@@ -137,6 +142,10 @@ jobs:
   # Restore dotnet CLI tools (e.g. pwsh, apicompat) before building.
   - template: /eng/pipelines/common/steps/restore-dotnet-tools.yml@self
 
+  # Download the assembly signing key for internal builds.
+  - ${{ if eq(parameters.isInternalBuild, true) }}:
+    - template: /eng/pipelines/common/steps/download-assembly-signing-key.yml@self
+
```
[the remark this answers is not in the document]
Comment (paulmedynski) [reply eng/pipelines/common/templates/jobs/ci-build-nugets-job.yml]: Fixed in 240ea7e57. Key download step in `ci-build-nugets-job.yml` is now gated on `ne(referenceType, 'Project')` in addition to `isInternalBuild`.

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlAuthenticationProviderManager.cs [hunk 18]
```diff
@@ -109,7 +109,7 @@ static SqlAuthenticationProviderManager()
...
```
[the remark this answers is not in the document]
Comment (paulmedynski) [reply src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlAuthenticationProviderManager.cs]: Fixed in 240ea7e57. Changed trace message from "assembly verification" to "strong-name identity verification".

File: eng/pipelines/common/steps/download-assembly-signing-key.yml
Comment (paulmedynski) [inline eng/pipelines/common/steps/download-assembly-signing-key.yml:1]: This will be a canditate to move into the new `shared/` directory when we start that relocation work.

File: eng/pipelines/jobs/pack-abstractions-package-ci-job.yml [hunk 19]
```diff
@@ -119,6 +124,23 @@ jobs:
...
 
+      # Build properties passed to dotnet pack.  Composed from a base set plus
+      # optional suffixes for Package-mode dependencies and assembly signing.
+      - name: baseBuildProperties
+        value: AbstractionsPackageVersion=${{ parameters.abstractionsPackageVersion }};AbstractionsAssemblyFileVersion=${{ parameters.abstractionsAssemblyFileVersion }}
+
+      # NOTE: We use compile-time ${{ if }} branches rather than concatenating
+      # separate variables (e.g. "$(base);$(optional);$(signing)") because
+      # when the optional variables are empty the semicolons remain, producing
+      # a trailing ";;" that MSBuild rejects with MSB1005.
```
Comment (paulmedynski) [inline eng/pipelines/jobs/pack-abstractions-package-ci-job.yml:135]: This is unfortunate MSBuild behaviour, so we must clutter things up here a bit.

File: eng/pipelines/onebranch/jobs/validate-signed-package-job.yml [hunk 20]
```diff
@@ -157,8 +157,8 @@ jobs:
           $nugetPackageInstallPath = "${{ variables.nugetPackageInstallPath }}"
           echo "nugetPackageInstallPath= $nugetPackageInstallPath"
 
-          # Verify strong name signing #####################################
-          echo "> 1. Verifying strong name signing of DLLs ..."
+          # Verify strong-name signing ###################################
```
Comment (paulmedynski) [inline eng/pipelines/onebranch/jobs/validate-signed-package-job.yml:160]: @paulmedynski I prefer 'assembly signing' since .NET doesn't do strong-name.

File: src/Microsoft.Data.SqlClient.Extensions/Abstractions/src/Abstractions.csproj [hunk 21]
```diff
@@ -79,6 +91,11 @@
                       Condition="'$(ReferenceType)' == 'Package'" />
   </ItemGroup>
 
+  <!-- Polyfills ======================================================= -->
+  <ItemGroup>
+    <PackageReference Include="System.Memory" />
```
Comment (paulmedynski) [inline src/Microsoft.Data.SqlClient.Extensions/Abstractions/src/Abstractions.csproj:96]: For `Span.SequenceEqual()` in .NET Standard.

File: src/Microsoft.Data.SqlClient.Extensions/Abstractions/src/SqlAuthenticationProvider.Internal.cs [hunk 22]
```diff
@@ -40,30 +40,75 @@ static Internal()
...
 
-            // If the MDS package is present, load its
+            // If the SqlClient assembly is present, load its
             // SqlAuthenticationProviderManager class and get/set methods.
             try
             {
-                // Try to load the MDS assembly.
+                // Try to load the SqlClient assembly.
+
+                #if ASSEMBLY_SIGNING
```
Comment (paulmedynski) [inline src/Microsoft.Data.SqlClient.Extensions/Abstractions/src/SqlAuthenticationProvider.Internal.cs:49]: The impetus for all of this work - verifying the SqlClient assembly here the same way SqlClient verifies the Azure assembly.

File: src/Microsoft.Data.SqlClient.Internal/Logging/test/SqlClientEventSourceTest.cs [hunk 23]
```diff
@@ -0,0 +1,20 @@
...
+// See the LICENSE file in the project root for more information.
+
+namespace Microsoft.Data.SqlClient.Internal.Logging.Test;
+
+public class SqlClientEventSourceTest
+{
+    [Fact]
+    public void SqlClientEventSource_Log_IsNotNull()
+    {
+        Assert.NotNull(SqlClientEventSource.Log);
```
Comment (paulmedynski) [inline src/Microsoft.Data.SqlClient.Internal/Logging/test/SqlClientEventSourceTest.cs:12]: Not the most exciting tests, but it's a start!

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlAuthenticationProviderManager.cs [hunk 24]
```diff
@@ -109,7 +109,7 @@ static SqlAuthenticationProviderManager()
                 SqlClientEventSource.Log.TryTraceEvent(
                     nameof(SqlAuthenticationProviderManager) +
                     $": Attempting to load Azure extension assembly={azureAssemblyName} without " +
-                    "strong name verification; ensure this assembly is from a trusted source");
+                    "strong-name identity verification; ensure this assembly is from a trusted source");
```
Comment (paulmedynski) [inline src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlAuthenticationProviderManager.cs:112]: @paulmedynski I prefer the ambiguous "assembly signing verification", so we can change what we verify without having to update the log message.

File: build.proj [hunk 25]
```diff
@@ -635,7 +645,12 @@
 
       <DotnetCommand>
         "$(DotnetPath)dotnet" test "$(SqlClientFunctionalTestProjectPath)"
+
+        <!-- Build arguments -->
```
Comment (paulmedynski) [inline build.proj:649]: Added consistent sections and ordering to these command builders.

**Comment by paulmedynski** [comment]
Moved this work to stack #4571 rooted at #4566.
